workspace/update-injected-packages/src/DirPatcher.ts

```diff
@@ -0,0 +1,208 @@
+
+  const added: AddedItem[] = newPaths
+    .filter(path => !(path in oldIndex))
+    .map(path => ({ path, newValue: newIndex[path] }))
+
+  const modified: ModifiedItem[] = oldPaths
+    .filter(path => path in newIndex && oldIndex[path] !== newIndex[path])
+    .map(path => ({ path, oldValue: oldIndex[path], newValue: newIndex[path] }))
+
+  return { added, removed, modified }
+}
+
+/**
+ * Apply a diff on a directory.
+ *
+ * The {@link optimizedDirDiff} is assumed to be already optimized (i.e. `removed` is already reversed).
+ */
+export async function applyDiff (optimizedDirDiff: DirDiff, sourceDir: string, targetDir: string): Promise<void> {
+  async function addRecursive (sourcePath: string, targetPath: string, value: Value): Promise<void> {
+    const makeParent = () => fs.mkdirSync(path.dirname(targetPath), { recursive: true })
+    if (value === DIR) {
+      await fs.promises.mkdir(targetPath, { recursive: true })
+    } else if (typeof value === 'string') {
+      makeParent()
+      const symlinkTarget = value
+      if (process.platform === 'win32') {
+        const symlinkRealTarget = path.resolve(sourceDir, symlinkTarget)
+        await fs.promises.symlink(symlinkRealTarget, targetPath, 'junction')
+      } else {
+        await fs.promises.symlink(symlinkTarget, targetPath)
+      }
+    } else if (typeof value === 'number') {
+      makeParent()
+      await fs.promises.link(sourcePath, targetPath)
+    } else {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const _: never = value // static type guard
+    }
+  }
+
+  async function removeRecursive (targetPath: string): Promise<void> {
+    try {
+      await fs.promises.rm(targetPath, { recursive: true, force: true })
+    } catch (error) {
+      if (!util.types.isNativeError(error) || !('code' in error) || (error.code !== 'ENOENT')) {
+        throw error
+      }
+    }
+  }
+
+  const adding = Promise.all(optimizedDirDiff.added.map(async item => {
+    const sourcePath = path.join(sourceDir, item.path)
+    const targetPath = path.join(targetDir, item.path)
+    await addRecursive(sourcePath, targetPath, item.newValue)
+  }))
+
+  const removing = Promise.all(optimizedDirDiff.removed.map(async item => {
+    const targetPath = path.join(targetDir, item.path)
+    await removeRecursive(targetPath)
+  }))
+
+  const modifying = Promise.all(optimizedDirDiff.modified.map(async item => {
+    const sourcePath = path.join(sourceDir, item.path)
+    const targetPath = path.join(targetDir, item.path)
+    if (item.oldValue === item.newValue) return
+    await removeRecursive(targetPath)
+    await addRecursive(sourcePath, targetPath, item.newValue)
+  }))
+
+  await Promise.all([adding, removing, modifying])
+}
+
+/**
+ * Convert a files map, which is a map from relative path of each file to their real paths,
+ * into an inodes map, which is a map from relative path of every file and directory to their inode type.
```

Comment:
What inode type? Isn't inode always a number?

Reply:
Inode is not a number. It is a data structure that stores attribute of a file or a directory (kind of like a fat pointer). So logically, an inode type is the type of the filesystem entry it represents: a file, a directory, a symlink, a mount point, a device, etc.

https://en.wikipedia.org/wiki/Inode

Anyway, I don't have a good name. If you can come up with something better, please give suggestion.